### PR TITLE
SIP 82 - generalised keeper system

### PR DESCRIPTION
**Pull request by jacko125**
When opening a pull request to submit a new SIP, please use the suggested template: https://github.com/Synthetixio/SIPs/blob/master/sip-X.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing WIP PRs.
 - The build passes.
 - Your Github username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

### CODE REVIEW

File: SIPS/sip-keeper-system.md [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+- Ensure keeper system to incentivise `Exchange.settle()` is called to settle fee reclamation in a timely manner.
+
+- Unsettled trades cause an in-determinate effect on the debt pool.
+
+SNX incentive to be paid out proportional to the `trade amount size and exchange fees` paid to make it not economically profitable to create multiple small exchanges for settlement to drain the incentives pool. The optimal balance would be that the USD value of the keeper rewards would be less than the `exchange fees paid + gas cost cost of an exchange transaction`, so this would discourage gaming of the keeper's rewards.
+
+| Action   | Gas Cost |
+|----------|----------|
+| Exchange | 405,865  |
+| Settle   | 51860    |
```
Comment (kaiynne) [inline SIPS/sip-keeper-system.md:67]: Insert comma, 51,860

File: SIPS/sip-keeper-system.md [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+title: Synthetix Keeper System
+status: WIP
+author: Jackson Chan (@jacko125)
+discussions-to: https://discord.gg/kPPKsPb
+
+created: 2020-07-14
+---
+
+## Simple Summary
+A generalised Synthetix keeper contract will be able to incentivise certain functions in the Synthetix protocol to be managed in a decentralised manner. The rewards, which can be SNX or other tokens / ether will be held in the keeper contract and the rewards schedule managed by the keeper contract.
```
Comment (justinjmoses) [inline SIPS/sip-keeper-system.md:12]: This is a bit long - can you shorten it?

File: SIPS/sip-keeper-system.md [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+## Abstract
+This SIP proposes a generalised Synthetix keeper contract to reward keeper functions like Exchange settlement, iSynth freezing, Chainlink pull request for prices and Binary option market resolutions.
+
+## Motivation
+The reason for a generalised keeper system / contract is to provide reward incentives for decentralised keepers to help manage the Synthetix protocol. The generalised keeper system will be able to set the amount of incentive and rewards for each function in the platform that requires keepers to manage.
+
+As the platform moves towards using decentralized oracle feeds from Chainlink, it will require keepers to manage the freezing and purging of iSynths (as per [SIP-61](./sip-61.md)).
+
+Another important function that requires keepers is to settle fee reclamation Exchange entries if they are not settled during the oridinary exchange process of a user's trading. Unsettled fee reclamations create an in-determinate state on the system's debt pool once they can be settled after the waiting period or next price is available.
```
Comment (justinjmoses) [inline SIPS/sip-keeper-system.md:22]: *indeterminate

File: SIPS/sip-keeper-system.md [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+Fee clamation can also be broken in the case that Chainlink's aggregators upgrade their contracts and the `roundID` values are reset, so that the efficient settlement of exchanges ensure that the debt pool is up to date and won't cause
```
Comment (justinjmoses) [inline SIPS/sip-keeper-system.md:24]: *reclamation

File: SIPS/sip-keeper-system.md [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+The reward incentives should cover the gas costs of executing the transaction in an efficient timely manner, ie using a 'fast' gas price to execute the transaction and be profitable for the keeper to continue operating.
```
Comment (justinjmoses) [inline SIPS/sip-keeper-system.md:26]: *i.e.

File: SIPS/sip-keeper-system.md [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+Keepers will play an important role in managing the Synthetix Futures system. The keepers will be incentivised to perform functions as required to maintain trader's positions and fee reclamation settlement for these positions.
+
+**Liquidations**
+
+A `Liquidation Fee` is paid to keepers, from the trader's maintenance margin, as a strong incentive to liquidate positions that have been liquidated. Synthetix futures positions that have gone below their liquidation price / maintenance margin would require keepers to liquidate them in a timely manner.
+
+The keeper system wouldn't be required to pay a reward to cover the cost of liquidating futures position.
+
+# Rationale
```
Comment (justinjmoses) [inline SIPS/sip-keeper-system.md:118]: empty?

File: SIPS/sip-keeper-system.md [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+
+Another important function that requires keepers is to settle fee reclamation Exchange entries if they are not settled during the oridinary exchange process of a user's trading. Unsettled fee reclamations create an in-determinate state on the system's debt pool once they can be settled after the waiting period or next price is available.
+
+Fee clamation can also be broken in the case that Chainlink's aggregators upgrade their contracts and the `roundID` values are reset, so that the efficient settlement of exchanges ensure that the debt pool is up to date and won't cause
+
+The reward incentives should cover the gas costs of executing the transaction in an efficient timely manner, ie using a 'fast' gas price to execute the transaction and be profitable for the keeper to continue operating.
+
+## Specification
+
+# Overview
```
Comment (justinjmoses) [inline SIPS/sip-keeper-system.md:30]: Please check the formatting of these headings - you've lost the hierarchy that is currently in the template

File: SIPS/sip-keeper-system.md [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+There are a number of core functions on Synthetix protocol that require keepers to perform and execute in a decentralised manner. In order to incentivise these keeper functions and cover the gas costs for the on-chain transactions, a generalised keeper rewards contract will provide reward incentives to keepers.
+
+### Actions
```
Comment (justinjmoses) [inline SIPS/sip-keeper-system.md:33]: Detailing these potential seems out of scope here - they may well change in the future (meaning this SIP will become stale) and frankly aren't critical to outline in detail for this SIP. I'd just summarize the current use-cases: `Synthetix.mint`, `Exchanger.settle` and `ExchangeRates.freeze` (TBD). That said, if you want to detail different features then it's fine to add a sentence or two, but be careful to not distract from the general feature you're proposing by going into depth about the various implementation reasons.

File: SIPS/sip-keeper-system.md (hunk 6 above)
Comment (jacko125) [reply SIPS/sip-keeper-system.md]: @justinjmoses  Have updated this section now

**Comment by jacko125** [comment]
> A few comments to address

@justinjmoses  Could you review changes and merge? Thanks

File: SIPS/sip-keeper-system.md [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+
+## Motivation
+The reason for a generalised keeper system / contract is to provide reward incentives for decentralised keepers to help manage the Synthetix protocol. The generalised keeper system will be able to set the amount of incentive and rewards for each function in the platform that requires keepers to manage.
+
+As the platform moves towards using decentralized oracle feeds from Chainlink, it will require keepers to manage the freezing and purging of iSynths (as per [SIP-61](./sip-61.md)).
+
+Another important function that requires keepers is to settle fee reclamation for exchange entries if they are not settled during the oridinary exchange process of a user's trading. Unsettled fee reclamations create an indeterminate state on the system's debt pool, once they can be settled after the waiting period or next price is available.
+
+Fee reclamation can also be broken in the case that Chainlink's aggregators upgrade their contracts and the `roundID` values are reset, so that the efficient settlement of exchanges ensure that the debt pool is up to date and won't cause
```
Comment (justinjmoses) [inline SIPS/sip-keeper-system.md:24]: end of sentence...

File: SIPS/sip-keeper-system.md (hunk 8 above)
Comment (justinjmoses) [reply SIPS/sip-keeper-system.md:34]: I still think these are out of scope and should simply be left as summaries

File: SIPS/sip-keeper-system.md [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+
+## Configurable Values (Via SCCP)
+
+Each action that a keeper could perform will have a configurable amount of USD value / SNX / token rewards that are paid to cover gas costs and running a keeper system.
+
+`gas: currency/eth rate` : 235.00 usd/eth
+
+| Action | Est Gas Cost | Gas Price | $USD Cost | $USD Reward |
+|--------|--------------|-----------|------------|------------|
+| 1. FreezeSynths for iSynths | 42,000| 35 gwei/gas | 0.34545 | 50.0 |
```
Comment (justinjmoses) [inline SIPS/sip-keeper-system.md:156]: I think this is too much detail for the SIP - the goal is to create a generalized keeper system, not to stipulate how much each should return.

**Pull request merged**